libsoftwarecontainer/src/container.cpp

```diff
@@ -622,12 +625,24 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
                 log_error() << "Could not create file " << tempPath;
                 return false;
             }
-            m_cleanupHandlers.push_back(new FileCleanUpHandler(tempPath));
         }
     }
 
-    return bindMountCore(pathInHost, pathInContainer, tempPath, readOnly);
+    if (!bindMountCore(pathInHost, pathInContainer, tempPath, readOnly)) {
+        // bindMountInContainer operation is not successful. Clean mounted tempPath.
+        MountCleanUpHandler rollback{tempPath};
+        rollback.clean();
+        return false;
+    }
 
+    // bindMountInContainer succeed. Add cleanup for mounted tempPath
+    m_createDirList.push_back(std::move(createDirInstance));
+    if (!pathIsDirectory) {
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
```

Comment:
You could probably rewrite this so that you use the same MountCleanupHandler as mentioned above (the one that cleans instantly), and then depending on result clean or save to `m_cleanupHandlers`.